tests/unit/socket_test.cc

```diff
@@ -546,3 +547,613 @@ SEASTAR_THREAD_TEST_CASE(inet_local_remote_address_sanity) {
+            header.push_back((src_in.sin_addr.s_addr >> 0) & 0xFF);
+            header.push_back((src_in.sin_addr.s_addr >> 8) & 0xFF);
+            header.push_back((src_in.sin_addr.s_addr >> 16) & 0xFF);
+            header.push_back((src_in.sin_addr.s_addr >> 24) & 0xFF);
+
+            // Destination address
+            header.push_back((dst_in.sin_addr.s_addr >> 0) & 0xFF);
+            header.push_back((dst_in.sin_addr.s_addr >> 8) & 0xFF);
+            header.push_back((dst_in.sin_addr.s_addr >> 16) & 0xFF);
+            header.push_back((dst_in.sin_addr.s_addr >> 24) & 0xFF);
+
+            // Source port
+            uint16_t src_port = ntohs(src_in.sin_port);
+            header.push_back((src_port >> 8) & 0xFF);
+            header.push_back((src_port >> 0) & 0xFF);
+
+            // Destination port
+            uint16_t dst_port = ntohs(dst_in.sin_port);
+            header.push_back((dst_port >> 8) & 0xFF);
+            header.push_back((dst_port >> 0) & 0xFF);
+        } else if (family_proto == 0x21) { // IPv6 TCP
+            addr_len = 36;
+            auto src_in6 = src_addr.as_posix_sockaddr_in6();
+            auto dst_in6 = dst_addr.as_posix_sockaddr_in6();
+
+            // Source address (16 bytes)
+            header.insert(header.end(),
+                reinterpret_cast<const char*>(&src_in6.sin6_addr),
+                reinterpret_cast<const char*>(&src_in6.sin6_addr) + 16);
+
+            // Destination address (16 bytes)
+            header.insert(header.end(),
+                reinterpret_cast<const char*>(&dst_in6.sin6_addr),
+                reinterpret_cast<const char*>(&dst_in6.sin6_addr) + 16);
+
+            // Source port
+            uint16_t src_port = ntohs(src_in6.sin6_port);
+            header.push_back((src_port >> 8) & 0xFF);
+            header.push_back((src_port >> 0) & 0xFF);
+
+            // Destination port
+            uint16_t dst_port = ntohs(dst_in6.sin6_port);
+            header.push_back((dst_port >> 8) & 0xFF);
+            header.push_back((dst_port >> 0) & 0xFF);
+        }
+    } else if ((version_cmd & 0x0F) == 0x00) { // LOCAL command
+        addr_len = 0;  // No address data for LOCAL
+    }
+
+    // Add TLVs
+    header.insert(header.end(), tlvs.begin(), tlvs.end());
+    addr_len += tlvs.size();
+
+    // Fill in length
+    header[len_pos] = (addr_len >> 8) & 0xFF;
+    header[len_pos + 1] = addr_len & 0xFF;
+
+    return header;
+}
+
+// Helper function for negative tests - expects connection to be rejected
+static void test_proxy_header_negative(
+    socket_address listen_addr,
+    std::vector<char> header) {
+
+    listen_options lo;
+    lo.reuse_address = true;
+    lo.proxy_protocol = true;
+
+    server_socket ss = seastar::listen(listen_addr, lo);
+
+    bool got_connection = false;
+
+    auto server = seastar::async([&ss, &got_connection] {
+        using namespace std::chrono_literals;
+        // Set a timer to abort accept after 100ms
```

Comment:
I don't understand this. I think we expect the connection to be rejected in a clean reliable way - never to need this timeout. Will this cause the test to fail - or incorrectly succeed - if the timeout occurs? Are we sure 100ms is always enough even on extremely overloaded test machines?

Reply:
The server never sees a bad connection. It's rejected by the networking layer. It's like a malformed SYN packet, an accept will never see it. So we need to escape.

Seastar tests aren't overcommitted.

I added checks that we get the timeout when we expect it.